fuse_models/include/fuse_models/common/throttled_callback.h

```diff
@@ -0,0 +1,159 @@
+
+#include <functional>
+
+
+namespace fuse_models
+{
+
+namespace common
+{
+
+/**
+ * @brief A throttled callback that encapsulates the logic to throttle a message callback so it is only called after a
+ * given period in seconds (or more). The dropped messages can optionally be received in a dropped callback, that could
+ * be used to count the number of messages dropped.
+ *
+ * @tparam M The message the callback receives
+ */
+template <class M>
+class ThrottledCallback
+{
+public:
+  using MessageConstPtr = typename M::ConstPtr;
+  using Callback = std::function<void(const MessageConstPtr&)>;
+
+  /**
+   * @brief Constructor
+   *
+   * @param[in] keep_callback   The callback to call when the message is kept, i.e. not dropped. Defaults to nullptr
+   * @param[in] drop_callback   The callback to call when the message is dropped because of the throttling. Defaults to
+   *                            nullptr
+   * @param[in] throttle_period The throttling period duration in seconds. Defaults to 0.0, i.e. no throttling
+   */
+  ThrottledCallback(Callback&& keep_callback = nullptr,  // NOLINT(whitespace/operators)
+                    Callback&& drop_callback = nullptr,  // NOLINT(whitespace/operators)
+                    const ros::Duration& throttle_period = ros::Duration(0.0))
+    : keep_callback_(keep_callback), drop_callback_(drop_callback), throttle_period_(throttle_period)
+  {
+  }
+
+  /**
+   * @brief Throttle period getter
+   *
+   * @return The current throttle period duration in seconds being used
+   */
+  const ros::Duration& getThrottlePeriod() const
+  {
+    return throttle_period_;
+  }
+
+  /**
+   * @brief Throttle period setter
+   *
+   * @param[in] throttle_period The new throttle period duration in seconds to use
+   */
+  void setThrottlePeriod(const ros::Duration& throttle_period)
+  {
+    throttle_period_ = throttle_period;
+  }
+
+  /**
+   * @brief Keep callback setter
+   *
+   * @param[in] keep_callback The new keep callback to use
+   */
+  void setKeepCallback(const Callback& keep_callback)
+  {
+    keep_callback_ = keep_callback;
+  }
+
+  /**
+   * @brief Drop callback setter
+   *
+   * @param[in] drop_callback The new drop callback to use
+   */
+  void setDropCallback(const Callback& drop_callback)
+  {
+    drop_callback_ = drop_callback;
+  }
+
+  /**
+   * @brief Last called time
+   *
+   * @return The last time the keep callback was called
+   */
+  const ros::Time& getLastCalledTime() const
+  {
+    return last_called_time_;
+  }
+
+  /**
+   * @brief Message callback that throttles the calls to the keep callback provided. When the message is dropped because
+   * of throttling, the drop callback is called instead.
+   *
+   * @param[in] message The input message
+   */
+  void callback(const MessageConstPtr& message)
+  {
+    const auto now = ros::Time::now();
+    if (!last_called_time_.isValid() || now - last_called_time_ > throttle_period_)
+    {
+      keep_callback_(message);
```

Comment:
The `keep_callback_`  could be null.

```suggestion
      if (keep_callback_) keep_callback_(message);
```

Reply:
Done